core/matrix/fft.cpp

```diff
@@ -0,0 +1,201 @@
+        auto dense_x = as<Dense<std::complex<float>>>(x);
+        get_executor()->run(fft::make_fft(float_b, dense_x, inverse_, buffer_));
+    } else {
+        auto dense_b = as<Dense<value_type>>(b);
+        auto dense_x = as<Dense<value_type>>(x);
+        get_executor()->run(fft::make_fft(dense_b, dense_x, inverse_, buffer_));
+    }
+}
+
+
+void Fft::apply_impl(const LinOp *alpha, const LinOp *b, const LinOp *beta,
+                     LinOp *x) const
+{
+    if (auto float_x = dynamic_cast<Dense<std::complex<float>> *>(x)) {
+        auto clone_x = x->clone();
+        this->apply_impl(b, lend(clone_x));
+        float_x->scale(beta);
+        float_x->add_scaled(alpha, lend(clone_x));
+    } else {
+        auto dense_x = as<Dense<value_type>>(x);
+        auto clone_x = x->clone();
+        this->apply_impl(b, lend(clone_x));
+        dense_x->scale(beta);
+        dense_x->add_scaled(alpha, lend(clone_x));
+    }
+}
+
+
+std::unique_ptr<LinOp> Fft2::transpose() const
+{
+    return Fft2::create(this->get_executor(), size1_, size2_, inverse_);
+}
+
+
+std::unique_ptr<LinOp> Fft2::conj_transpose() const
+{
+    return Fft2::create(this->get_executor(), size1_, size2_, !inverse_);
+}
+
+
+bool Fft2::is_inverse() const { return inverse_; }
+
+
+void Fft2::apply_impl(const LinOp *b, LinOp *x) const
+{
+    if (auto float_b = dynamic_cast<const Dense<std::complex<float>> *>(b)) {
+        auto dense_x = as<Dense<std::complex<float>>>(x);
+        get_executor()->run(fft::make_fft2(float_b, dense_x, size1_, size2_,
+                                           inverse_, buffer_));
+    } else {
+        auto dense_b = as<Dense<value_type>>(b);
+        auto dense_x = as<Dense<value_type>>(x);
+        get_executor()->run(fft::make_fft2(dense_b, dense_x, size1_, size2_,
+                                           inverse_, buffer_));
+    }
+}
+
+
+void Fft2::apply_impl(const LinOp *alpha, const LinOp *b, const LinOp *beta,
+                      LinOp *x) const
+{
+    if (auto float_x = dynamic_cast<Dense<std::complex<float>> *>(x)) {
+        auto clone_x = x->clone();
+        this->apply_impl(b, lend(clone_x));
+        float_x->scale(beta);
+        float_x->add_scaled(alpha, lend(clone_x));
+    } else {
+        auto dense_x = as<Dense<value_type>>(x);
+        auto clone_x = x->clone();
+        this->apply_impl(b, lend(clone_x));
+        dense_x->scale(beta);
+        dense_x->add_scaled(alpha, lend(clone_x));
+    }
+}
+
+
+std::unique_ptr<LinOp> Fft3::transpose() const
+{
+    return Fft3::create(this->get_executor(), size1_, size2_, size3_, inverse_);
+}
+
+
+std::unique_ptr<LinOp> Fft3::conj_transpose() const
+{
+    return Fft3::create(this->get_executor(), size1_, size2_, size3_,
+                        !inverse_);
+}
+
+
+bool Fft3::is_inverse() const { return inverse_; }
+
+
+void Fft3::apply_impl(const LinOp *b, LinOp *x) const
+{
+    if (auto float_b = dynamic_cast<const Dense<std::complex<float>> *>(b)) {
```

Comment:
Why do you only special the complex float not others?
Does Fft only support complex Dense now?

Reply:
I wasn't familiar enough with the symmetries in R2C and C2R FFT to implement them in-place, so I left the reference implementation out for later. But I will probably address this soon before merging, yes.

Reply:
I remembered that R2C and C2R FFTs are rectangular operations (n x 2n and 2n x n, respectively), so they don't fit into this scheme right now.

Reply:
The complex-real case is only important for real LinOps. since FFT is complex, we don't need anything related to complex IDR shadow spaces.